Survey Component

trm313 opened the pull request:
As I worked through, I realized that I might have to rework the data scheme, and the overall logic of the component.. 
Before I dive into that I wanted to make sure I'm conceptually on the right page as far as layout

For now, please review (ignoring shitty / non-existent styling) for the flow of the survey, and the general layout and process

Neoflash1979 commented:
@trm313 As far as the general flow and layout, I think it's great. I'm sure it will look awesome once you start working your CSS magic. 

I'm a bit concerned about the way the questions are formulated. If you look at the conditions of satisfaction for the related user stories, we were thinking of approaching the survey with a future perspective "What would you like to work with?" whereas you have formulated your questions thinking about the past or present "What do you work with now?"

Since our principal user persona is someone that is quite possibly looking to gain some experience working with a new technology I feel that the questions we ask in the first part of the survey should make clear that we are interested in learning about what they would like to develop in, not just what they have developed in so far in their programming journey. Does that make sense? Did you have a different vision?

trm313 commented:
Thanks for those tips Phil - I agree that it needs major cleaning up / reduce duplication, and refactor to leverage reusable code for it. Just completed some styling / inclusion of the Devicons, and now I'm working on refactoring, slowly but surely.

Neoflash1979 commented:
@trm313 I normally would not fool around in one of your branches but I've take the liberty to push a simple fix to the way you were including devicon in the project. It was causing a build error.

I've noticed in the package.json that you have included devicon and devicon-2.2. Do we need both? If you are sure that we won't be using one of the two, you can go ahead and delete the corresponding entry from package.json.

trm313 commented:
Thanks for taking care of that Phil - I merged the changes into my branch. 

Additional updates:
- Refactored selection functions to use a single function
- Refactored languages / frameworks / backends pages to be generated from a single HTML block, that dynamically receives data for each
- Standardized the styling to allow for templatized pages

Next Steps:
- Add dynamic navigation to the new templatized HTML components
- - Could handle navigation inside the 'navigateToPage()' function, and simply send 'next' and 'back'. Coupled with currentPage state, should be able to direct okay
- With templatization, the "Languages" page headers (devices eg. Web, Gaming) lost their images since they don't use Devicons. Need to find a way to keep it there
- Ratings page:
- - Make the dots fill up when hovered, but don't send until clicked
- Complete styling
- Responsive styling

Neoflash1979 commented on src/app/survey/survey.component.ts:
```diff
@@ -9,53 +11,161 @@ export class SurveyComponent implements OnInit {
-  data = dataSchema;
+  data = technologyData;
 
   response = {
```
To make sure that the response object adheres to the `SurveyResponse` interface you should do this: 
`response: SurveyResponse = { `

Also, you should make sure that what you are saving to localStorage adheres to the `Survey` class. It needs to have the ìsCompleted` property so that a returning user having completed the survey won't be shown the survey and instead be directed to the Match List.

Both of these can be imported from `survey.model.ts`.

trm313 commented:
COMPLETED
- [X] We should probably save the survey state to localStorage after each step of the survey, not only at the end, and make it so that when users come back the survey starts where they left off, if ever they only complete part of the survey. (Low priority)
- [X] I wonder if there is an easy way to make the box the same size for each survey section. (Low priority)
- [X] If you decide to leave refactoring the code to later, I would suggest documenting your code a bit more in the meantime. It's not all that easy to understand what is going on. You could add more comments or even a comment block at the top that gives an overview of how everything flows. (Medium priority)

REMAINING
- [ ] Rating dots: It's cool that the dot under the mouse lights up. It would be even better if all the dots before it would also light up. (Low priority)
- [ ] Although you have refactored your code, I still think it would benefit from some more aggressive abstracting/refactoring. I'm being picky here, you did a great job of implementing the functionalities we need. But I'm thinking that if we ever need to add more questions or options to the survey, it ought to be easier to do so.
- [ ] All the survey pages have many elements in common: a box, a title, rows of possible answers and navigation buttons. I'm thinking that your template should only have that one skeleton and that the content within these elements should be generated through the component. If it's not possible, then it might be a good idea to have children component instead for each of the survey section, in order to clean things up and modularize them. (Medium priority)

- [ ] This PR partially meets the conditions of satisfaction of Issue #20. It's not a reason to hold it back as it completes the other survey related issues #12, #14 and #21.

Neoflash1979 commented:
@trm313 Found a little something that needs to be fixed. It shouldn't be possible to hit the "next" arrow button on each survey page without having first fully answered the question.

trm313 commented:
That's a good point, and I agree for "Devices" and "Languages". 

However, I think "Frameworks" and "Backends" should be the exception, and allow the user to progress without a selection (it's possible they simply haven't used either, just straight up PHP, for example)

trm313 commented:
The checks have passed for the latest commits which include sending the Survey response on user signup. Is there a way I can access the deployed MongoDb to confirm this is sending correctly?
